### PR TITLE
feat: conform to LSP semantic token specification

### LINKED ISSUE #2483: Semantic tokens do not conform to the LSP specification
Semantic highlighting was recently merged into Neovim. Unfortunately the token types provided by Omnisharp don't conform to the [LSP semantic tokens specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) so Neovim cannot support it (neovim/neovim#21391).

To demonstrate the issue, here's a comparison of Lua Language Server's tokens to Omnisharp's tokens. Omnisharp's tokens are fundamentally wrong as well as having illegal formatting.

<details>
  <summary>Lua Language Server (correct tokens)</summary>

  ``` lua
{
   full = true,
   legend = {
      tokenModifiers = {
         "declaration",
         "definition",
         "readonly",
         "static",
         "deprecated",
         "abstract",
         "async",
         "modification",
         "documentation",
         "defaultLibrary",
         "global",
      },
      tokenTypes = {
         "namespace",
         "type",
         "class",
         "enum",
         "interface",
         "struct",
         "typeParameter",
         "parameter",
         "variable",
         "property",
         "enumMember",
         "event",
         "function",
         "method",
         "macro",
         "keyword",
         "modifier",
         "comment",
         "string",
         "number",
         "regexp",
         "operator",
         "decorator",
      },
   },
   range = true,
}
  ```

</details>

<details>
  <summary>Omnisharp-roslyn (incorrect Tokens)</summary>

  ``` lua
{
   full = vim.empty_dict(),
   legend = {
      tokenModifiers = { "static symbol" },
      tokenTypes = {
         "comment",
         "excluded code",
         "identifier",
         "keyword",
         "keyword - control",
         "number",
         "operator",
         "operator - overloaded",
         "preprocessor keyword",
         "string",
         "whitespace",
         "text",
         "static symbol",
         "preprocessor text",
         "punctuation",
         "string - verbatim",
         "string - escape character",
         "class name",
         "delegate name",
         "enum name",
         "interface name",
         "module name",
         "struct name",
         "type parameter name",
         "field name",
         "enum member name",
         "constant name",
         "local name",
         "parameter name",
         "method name",
         "extension method name",
         "property name",
         "event name",
         "namespace name",
         "label name",
         "xml doc comment - attribute name",
         "xml doc comment - attribute quotes",
         "xml doc comment - attribute value",
         "xml doc comment - cdata section",
         "xml doc comment - comment",
         "xml doc comment - delimiter",
         "xml doc comment - entity reference",
         "xml doc comment - name",
         "xml doc comment - processing instruction",
         "xml doc comment - text",
         "xml literal - attribute name",
         "xml literal - attribute quotes",
         "xml literal - attribute value",
         "xml literal - cdata section",
         "xml literal - comment",
         "xml literal - delimiter",
         "xml literal - embedded expression",
         "xml literal - entity reference",
         "xml literal - name",
         "xml literal - processing instruction",
         "xml literal - text",
         "regex - comment",
         "regex - character class",
         "regex - anchor",
         "regex - quantifier",
         "regex - grouping",
         "regex - alternation",
         "regex - text",
         "regex - self escaped character",
         "regex - other escape",
      },
   },
   range = true,
}
  ```

</details>

### PR DESCRIPTION
**Pull request by amaanq**
Closes #2483 hopefully. I would need help figuring out how to implement nested captures, like variable.readonly, but this is a good start imo

### CODE REVIEW

**Comment by amaanq** [comment]
Ping @dibarbet, sorry but I'd like this in as soon as possible... :)

**Comment by tony-dwire-paradigm** [comment]
@amaanq thanks for putting in this work, I was very surprised to see this issue after updating omnisharp

**Comment by misaelaguayo** [comment]
Any updates on this? Would be great to merge

**Comment by srodrigo** [comment]
Any updates on this PR? :pray:

**Comment by gautaz** [comment]
Hello,

As I am also interested in seeing this PR merged, I had a look at the `CONTRIBUTING.md` document available at the root of the project, this document states:

> Before contributing any feature, make sure to discuss it upfront with the team - this can be done on Github in an issue or on Slack.

What is unclear to me is who is "the team".
@filipw As you seem to be the author of this document, would you perhaps give us an insight on this matter?
My apologies for barging in this way but I am just wondering how contributions are elected for merging.
I can also open a new issue to discuss this specific matter if necessary.

**Comment by LucasTavaresA** [comment]
i think this is not possible cause most of those are C# reserved keywords

File: tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs [hunk 1]
```diff
@@ -7,178 +7,146 @@
...
-        {
-        }
-
-        protected override string EndpointName => OmniSharpEndpoints.Highlight;
-
-        [Fact]
-        public async Task HighlightSingleLine()
-        {
-            var testFile = new TestFile("a.cs", @"
+namespace OmniSharp.Roslyn.CSharp.Tests {
```
Comment (CEbbinghaus) [inline tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs:10]: It seems the formatting in this file is a bit broken. The braces should be on the next line

File: tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs [hunk 2]
```diff
@@ -7,178 +7,146 @@
...
-                Punctuation(";"),
-                Punctuation("}"));
-        }
+    var highlights = await GetHighlightsAsync(testFile, line: 3);
 
-        [Fact]
-        public async Task HighlightEntireFile()
-        {
-            var testFile = new TestFile("a.cs", @"
+    AssertSyntax(highlights, testFile.Content.Code, 3, Keyword("class"),
```
Comment (CEbbinghaus) [inline tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs:30]: Again formatting. Its clearer when all the arguments are on newlines

**Review by CEbbinghaus — CHANGES_REQUESTED** [review]
Formatting is off for the HighlightFacts. Otherwise lgtm.
Not sure how those tokens get turned into the text getting sent to the server so I wouldn't know if this fixes the problem itself.

File: tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs [hunk 3]
```diff
@@ -7,178 +7,146 @@
 using Xunit;
```
Comment (JoeRobich) [inline tests/OmniSharp.Roslyn.CSharp.Tests/HighlightFacts.cs:7]: Changes to this file should be reverted. The OmniSharp v1 Highlight service is not intended to drive SemanticHighlight. It may still be used by some Editor extensions.

File: src/OmniSharp.Abstractions/Models/v2/SemanticHighlight/SemanticHighlightClassification.cs [hunk 4]
```diff
@@ -20,43 +20,42 @@ public enum SemanticHighlightClassification
...
+        XmlLiteralAttribute,
         XmlLiteralAttributeQuotes,
         XmlLiteralAttributeValue,
         XmlLiteralCDataSection,
         XmlLiteralComment,
         XmlLiteralDelimiter,
         XmlLiteralEmbeddedExpression,
         XmlLiteralEntityReference,
-        XmlLiteralName,
+        XmlLiteral,
```
Comment (JoeRobich) [inline src/OmniSharp.Abstractions/Models/v2/SemanticHighlight/SemanticHighlightClassification.cs:58]: Please revert the changes to the Xml DocComment and Literal classification names. They are taken straight from Roslyn. 
https://github.com/dotnet/roslyn/blob/59727ee4103c75edb799bb46e081427a2e1940ef/src/Workspaces/Core/Portable/Classification/ClassificationTypeNames.cs#L55-L73

**Review by JoeRobich — COMMENTED** [review]
Using the built-in tokens where possible is a good change. Are there other changes that need to be made to conform to the token spec?

**Comment by waynebowie99** [comment]
@amaanq Is this still in the works? Just ran into this issue after up this pull request was completed neovim/nvim-lspconfig#2692 and I upgraded to nvim 0.9.

**Comment by 333fred** [comment]
Closing out in favor of https://github.com/OmniSharp/omnisharp-roslyn/pull/2548
